Product Collection: Add stock status filter

imanish003 opened the pull request:
<!-- Start by describing the changes made in this Pull Request, and the reason for such changes. -->
This PR adds `Stock status` control to the Product collection block. 
<img width="267" alt="image" src="https://github.com/woocommerce/woocommerce-blocks/assets/16707866/a84b4dcf-caf6-4292-8636-de2f1a139365">


<!-- Reference any related issues or PRs here -->

Fixes #9362 

### Testing

1. Create a new post
2. Add Product Collection block
3. With the `Product Collection` block selected, check the sidebar for the `Stock status` control. Ensure that it is displayed as shown below:
<img width="267" alt="image" src="https://github.com/woocommerce/woocommerce-blocks/assets/16707866/a84b4dcf-caf6-4292-8636-de2f1a139365">

4. Test the functionality of the stock status filter by modifying its selection. Observe that the changes should immediately reflect in the Editor view. For instance, if `On backorder` is the selected option, only products with the `On backorder` status should be visible in the Editor.
5. Save and publish the post.
    - Navigate to the live post on the Frontend and verify that the changes you made in the Editor are accurately reflected.
6. Go back to the Editor and confirm that the filters can be reset using the `Reset` or `Reset all` button, as shown below:
<img width="279" alt="image" src="https://github.com/woocommerce/woocommerce-blocks/assets/16707866/dbea8bfc-2d6c-4ece-9236-4b8bf176f5e3">

* [ ] Do not include in the Testing Notes <!-- Check this box if this PR can't be tested by users (ie: it doesn't include user-facing changes or it can't be tested without manually modifying the code). -->

### WooCommerce Visibility

<!-- Check this [this doc](../docs/blocks/feature-flags-and-experimental-interfaces.md) to see if the change is visible in WC core or not (part of the feature plugin or experimental)-->

* [ ] WooCommerce Core
* [ ] Feature plugin
* [X] Experimental

### Changelog

> Product Collection: Add stock status control to sidebar settings

imanish003 committed:
Add columns control to product collection block editor settings

- `InspectorControls` from './inspector-controls' is now imported in `edit.tsx` and used in the returned JSX of `Edit` function.
- A new file `columns-control.tsx` is added under 'product-collection' block's 'inspector-controls' directory which exports a `ColumnsControl` component. This component uses `RangeControl` from '@wordpress/components' to control the number of columns in the product collection display layout when the layout type is 'flex'.
- The types file (`types.ts`) for 'product-collection' block is updated. The `Attributes` interface is renamed to `ProductCollectionAttributes` and the `ProductCollectionContext` interface is removed. The `ProductCollectionAttributes` now includes 'queryContext', 'templateSlug', and 'displayLayout' properties.

imanish003 committed:
Merge branch 'trunk' into 9356-product-collection-editor-settings-columns

imanish003 committed:
Refactor: Simplify Fallback Return in ColumnsControl Component

This commit simplifies the fallback return value of the ColumnsControl component. Instead of returning an empty fragment (<> </>), it now returns null when the condition isn't met. This change improves readability and aligns with best practices for conditional rendering in React.

imanish003 committed:
Feature: Add 'Order By' Control to Product Collection Inspector

This commit adds a new 'Order By' control to the product collection inspector. The control allows users to specify the order of products in a collection by various attributes such as title and date. To support this, a new component 'OrderByControl' has been created and included in the product collection inspector. Additionally, the types for 'order' and 'orderBy' attributes have been updated and exported for reuse.

imanish003 committed:
Merge branch 'trunk' of https://github.com/woocommerce/woocommerce-blocks into 9359-product-collection-editor-settings-order-by

imanish003 committed:
Add more options to OrderBy type

imanish003 committed:
Merge branch 'trunk' into 9359-product-collection-editor-settings-order-by

imanish003 committed:
Add orderby handling on frontend & editor

The main changes include:
1. Added a new property 'isProductCollectionBlock' in the block.json to denote if a block is a product collection block.
2. In the ProductCollection PHP class, a new initialization function has been defined to hook into the WordPress lifecycle, register the block, and update the query based on this block.
3. Added methods to manage query parameters for both frontend rendering and the Editor.
4. Expanded allowed 'collection_params' for the REST API to include custom 'orderby' values.
5. Defined a function to build the query based on block attributes, filters, and global WP_Query.
6. Created utility functions to handle complex query operations such as merging queries, handling custom sort values, and merging arrays recursively.

These improvements allow for more flexible and robust handling of product collections in both the front-end display and the WordPress editor. It also extends support for custom 'orderby' values in the REST API, which allows for more advanced sorting options in product collections.

imanish003 committed:
Merge branch 'trunk' into 9359-product-collection-editor-settings-order-by

imanish003 committed:
Merge branch 'trunk' into 9359-product-collection-editor-settings-order-by

imanish003 committed:
Merge branch 'trunk' of https://github.com/woocommerce/woocommerce-blocks into 9361-product-collection-filters-on-sale

imanish003 committed:
Add 'on sale' filter and enhance settings management in product collection block

This commit introduces several changes to the product collection block.
- First, it adds a new 'on sale' filter that can be used to display only the products that are currently on sale.
- It also refactors the settings management in the product collection block to use the experimental ToolsPanel component from WordPress, which provides a more flexible and intuitive way to manage block settings.
- It moves the 'Columns' control into the ToolsPanel, along with the 'Order by' control.
- A new utility function `setQueryAttribute` is introduced to simplify setting nested query parameters.
- The structure of the `ProductCollectionAttributes` and `ProductCollectionQuery` types have been adjusted to accommodate the changes.
- Finally, it makes corresponding changes in the PHP part to handle the new 'on sale' query parameter.

This should enhance the flexibility and user-friendliness of the product collection block.

imanish003 committed:
Merge branch 'trunk' into 9361-product-collection-filters-on-sale

imanish003 committed:
Add stock status filter to WooCommerce product collection block

This commit introduces a stock status filter to the WooCommerce product collection block.

The changes include:
1. Added the ability to filter products based on their stock status within the 'product-collection' block. A new stock status control is created within the inspector-controls of the block.
2. A new 'get_stock_status_query' function is introduced in 'ProductCollection.php' which returns a query for products depending on their stock status.

Please note that the stock status filter will only appear in the experimental build for now.

imanish003 committed:
Refactor Stock Status control of Product Collection block

This commit refactors the Stock Status control. The changes aim to improve the code organization and make the behavior of the component more explicit.

The key modifications are:
1. Moved stock status related constants and functions from `inspector-controls/utils.tsx` to `inspector-controls/constants.ts`. This is done to ensure that all constants and similar utility functions are organized in one place.
2. Updated `product-collection/index.tsx` to import `getDefaultStockStatuses` from `inspector-controls/constants` instead of `inspector-controls/utils`.
3. Updated `stock-status-control.tsx` to determine whether the stock status has value or not by comparing with the default stock statuses using `fastDeepEqual`. If the stock status control is deselected, it resets the stock status to the default statuses.

These changes do not introduce any new functionalities, but improve the readability and maintainability of the code.

imanish003 committed:
Merge branch 'trunk' into 9362-product-collection-filters-stock-status

imanish003 committed:
Resolve conflicts

Aljullu commented on src/BlockTypes/ProductCollection.php:
```diff
@@ -326,4 +351,94 @@ private function array_merge_recursive_replace_non_array_properties( $base, $new
+	 * @param array $stock_statii An array of acceptable stock statii.
+	 * @return array
+	 */
+	private function get_stock_status_query( $stock_statii ) {
```
TIL about _statii_, I guess that's the plural of _status_, no? I think in the codebase we use `statuses` more often than `statii`, so I would rename this variable to `$stock_statuses` if possible to keep it consistent. Not a strong opinion, though.

imanish003 replied:
Hey @Aljullu, I agree with you. I have updated all instances of `statii` in commit [0c30f5b](https://github.com/woocommerce/woocommerce-blocks/pull/9580/commits/0c30f5b51c4594654ffe5dbb0e937f83f623bde8). Thank you! 🙏🏻

Aljullu commented on src/BlockTypes/ProductCollection.php:
```diff
@@ -76,46 +76,54 @@ public function update_rest_query( $args, $request ): array {
-		$params['orderby']['enum'] = array_unique( array_merge( $original_enum, $this->custom_order_opts ) );
-		return $params;
+	private function get_final_query_args( $common_query_values, $query ) {
+		$orderby_query    = $query['orderby'] ? $this->get_custom_orderby_query( $query['orderby'] ) : [];
```
This causes a PHP warning in the frontend if `orderby` is not defined:

```
Warning: Undefined array key "orderby" in /wp-content/plugins/woocommerce-blocks/src/BlockTypes/ProductCollection.php on line 100
```

imanish003 replied:
Hey @Aljullu, Thanks for raising this 🙏🏻  I have fixed it in [d1e29ef](https://github.com/woocommerce/woocommerce-blocks/pull/9580/commits/d1e29ef02af1acc38614998a11a5a746eef48693) 🤝

Aljullu commented on src/BlockTypes/ProductCollection.php:
```diff
@@ -326,4 +351,94 @@ private function array_merge_recursive_replace_non_array_properties( $base, $new
+		}
+
+		/**
+		 * If all stock statuses are selected except 'outofstock', we use the
+		 * product visibility query to filter out out of stock products.
```
I have to say that, as a user, I would find this a bit confusing because the block settings and WC settings can contradict each other. IMO block settings should take priority over WooCommerce settings. So if the _Hide out of stock items from the catalog_ option is checked but I add a Product Collection block selecting to display products with all stock statuses, I would expect all products to appear, ignoring the _Hide out of stock items from the catalog_ option.

In other words, I think the _Hide out of stock items from the catalog_ option should affect which stock status options are selected by default when inserting the block, but they shouldn't affect how the block works once the block attributes have been set.

But... I don't have a strong opinion so I'm happy to change my mind. Also, I don't think it's a blocker.

imanish003 replied:
Hey @Aljullu,

> I think the Hide out of stock items from the catalog option should affect which stock status options are selected by default when inserting the block, but they shouldn't affect how the block works once the block attributes have been set.

I just want to confirm that currently, when we insert the block, we do take the "Hide out of stock items from the catalog" setting into consideration. This is implemented [here](https://github.com/woocommerce/woocommerce-blocks/blob/0c30f5b51c4594654ffe5dbb0e937f83f623bde8/assets/js/blocks/product-collection/constants.ts#L18-L32).

Also, I agree that it makes sense to ignore the "Hide out of stock items from the catalog" setting if we allow merchants to explicitly set stock statuses in the Product Collection block.
![image](https://github.com/woocommerce/woocommerce-blocks/assets/16707866/166961ab-ffd8-431c-aed9-92b90fbf3f5b)


PS: I will go ahead and merge this PR for now. We can continue our discussion here. If we decide to ignore the global "Hide out of stock items from the catalog" setting, I will create a new PR for that.

CC: @nerrad @tjcafferkey

Aljullu replied:
> I just want to confirm that currently, when we insert the block, we do take the "Hide out of stock items from the catalog" setting into consideration. This is implemented [here](https://github.com/woocommerce/woocommerce-blocks/blob/0c30f5b51c4594654ffe5dbb0e937f83f623bde8/assets/js/blocks/product-collection/constants.ts#L18-L32).

Oh right, I missed it! That's nice.

Aljullu reviewed (approved):
Code looks good and it's working as expected. I left some comments inline, but only the PHP warning is a blocking one IMO, so pre-approving (but feel free to ask for a re-review if you want).

imanish003 committed:
Fix: Default values of attributes not saving as serialized block comment

This was happening because of this issue: https://github.com/WordPress/gutenberg/issues/7342

Therefore, I had to use `useEffect` to set the default values of the attributes.

Here is the list of changes I made:
1. Removed default values from `block.json` for `query`, `tagName`, and `displayLayout`.
2. Moved the default values to `constants.ts` and created a new object `DEFAULT_ATTRIBUTES` to store them.
3. Relocated `constants.ts` from `inspector-controls` to the parent directory.
4. Refactored `edit.tsx` to use `DEFAULT_ATTRIBUTES` from `constants.ts` to set default attributes using `useEffect`.
5. Removed the attributes assignment from `registerBlockType` in `index.tsx`.
6. Updated `columns-control.tsx`, `index.tsx`, `order-by-control.tsx`, and `stock-status-control.tsx` to import from the relocated `constants.ts`.
7. Updated `ProductCollectionAttributes` and `ProductCollectionQuery` in `types.ts` to include `tagName` and `isProductCollectionBlock` respectively.
8. Modified `ProductCollection.php` to match the updated `orderBy` key in the query parameter.

This refactor enhances the readability of the code and reduces duplication by keeping all constants and default values in one place.

imanish003 committed:
Replace usage of 'statii' with 'statuses' in stock status handling

This commit replaces all instances of 'statii' with the correct term 'statuses' in the context of handling stock status. This change affects three files:
1. `assets/js/blocks/product-collection/inspector-controls/stock-status-control.tsx` - The term is corrected in a comment block.
2. `assets/js/blocks/product-collection/types.ts` - Updated the name of a variable in the `ProductCollectionQuery` interface.
3. `src/BlockTypes/ProductCollection.php` - Here, the term is replaced in several locations including variable names, comments and the method `get_stock_status_query`.

This commit helps improve code readability and consistency across the repository.

imanish003 committed:
Merge branch 'trunk' into 9362-product-collection-filters-stock-status
